src/runner.py

```diff
@@ -88,596 +47,416 @@ class LxdInstanceConfig(TypedDict):
-                    runner.local.stop(wait=True)
-                    try:
-                        runner.local.delete(wait=True)
-                    except Exception:  # nosec B110
-                        # Ephemeral containers should auto-delete when stopped;
-                        # this is just a fall-back.
-                        pass
-                else:
-                    # We somehow have a non-running instance which should have been
-                    # ephemeral. Try to delete it and allow any errors doing so to
-                    # surface.
-                    runner.local.delete(wait=True)
-            except Exception as e:
-                raise RunnerRemoveError(f"Failed remove local runner: {runner.name}") from e
-
-        # remove profile
-        try:
-            profile = self._lxd.profiles.get(runner.name)
-            profile.delete()
-        except (LXDAPIException, NotFound):
-            pass
+        if self.instance is None:
+            return
 
-    def _clean_unused_profiles(self) -> None:
-        """Clean all unused profiles created with this manager."""
-        for profile in self._lxd.profiles.all():
-            if profile.name.startswith(self.app_name) and not profile.used_by:
-                profile.delete()
+        # The LXD instance is meant to run untrusted workload. Hardcoding the tmp directory should
+        # be fine.
+        binary_path = "/tmp/runner.tgz"  # nosec B108
 
-    def _register_runner(self, instance: pylxd.models.Instance, labels: Sequence[str]) -> None:
-        """Register a runner in an instance.
+        logger.info("Installing runner binary on LXD instance: %s", self.name)
 
-        Args:
-            instance: LXD instance of the runner.
-            labels: Sequence of labels to tag the runner with.
-        """
-        api = self._api
-        if "/" in self.path:
-            owner, repo = self.path.split("/")
-            token = api.actions.create_registration_token_for_repo(owner=owner, repo=repo)
-        else:
-            token = api.actions.create_registration_token_for_org(org=self.path)
-        cmd = [
-            "sudo -u ubuntu /opt/github-runner/config.sh",
-            f"--url https://github.com/{self.path} ",
-            "--token ",
-            f"{token.token} ",
-            "--ephemeral ",
-            "--name ",
-            f"{instance.name} ",
-            "--unattended ",
-            "--labels ",
-            f"{','.join(labels)}",
-        ]
-        self._check_output(instance, cmd)
-
-    def _start_runner(self, instance: pylxd.models.Instance) -> None:
-        """Start a runner that is already registered.
+        # Creating directory and putting the file are idempotent, and can be retried.
+        self.instance.files.mk_dir(str(self.runner_application))
+        self.instance.files.put(binary_path, binary.read_bytes())
+        self._execute(["tar", "-xzf", binary_path, "-C", str(self.runner_application)])
```

Comment:
No risk of injecting another command in runner_application using `&& runner_application`

Reply:
The `self.runner_application` and `binary_path` are both hardcoded values.